File: tests/journal.lib/channel_default_device.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the journal
#include <pyre/journal.h>
// support
#include <cassert>


// type aliases
template <typename severityT>
using channel_t = pyre::journal::channel_t<severityT>;


// severity stub
class severity_t : public channel_t<severity_t>
{
    // metamethods
public:
    // index initialization is required...
    inline severity_t(const name_type & name): channel_t<severity_t>(name) {}
};


// the trash can
using trash_t = pyre::journal::trash_t;


// verify that we can control the default device
int main() {
    // get the default device
    auto builtin = severity_t::index().device();

    // make a new device
    auto custom = std::make_shared<trash_t>();
    // install it
    severity_t::index().device(custom);

    // check that the current device is the one we just installed
    assert (severity_t::index().device() == custom);

    // all done
    return 0;
}


// end of file

File: tests/journal.pkg/channel_default_device.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


# verify that we can control the default device
def test():
    """
    Verify that we can control the default device
    """
    # the trash can
    from journal.Trash import Trash as trash
    # get the channel
    from journal.Channel import Channel as channel

    # ask it for the default device
    builtin = channel.getDefaultDevice()

    # make new device
    custom = trash()
    # install it
    old = channel.setDefaultDevice(device=custom)

    # check that the default device is what we just installed
    assert channel.getDefaultDevice() is custom
    # and the device we just replaced was the original built-in one
    assert old is builtin

    # all done
    return


# main
if __name__ == "__main__":
    # prohibit the journal bindings
    journal_no_libjournal = True
    # run the test
    test()


# end of file
